feat: add the zig_shared_library rule to generate shared objects or dynamic libraries

Titousensei opened the pull request:
Solves #102

This is what I have so far. I'm not sure what command to run for the cross-platform tests. @aherrmann please advise.

aherrmann commented on zig/tests/target_platform_test.bzl:
```diff
@@ -112,6 +115,12 @@ def target_platform_test_suite(name):
         partial.make(_lib_file_extension_x86_64_linux_test, target_under_test = "//zig/tests/simple-library:library"),
         partial.make(_lib_file_extension_x86_64_windows_test, target_under_test = "//zig/tests/simple-library:library"),
+        # Test Zig target plaform on a shared library target
+        partial.make(_build_shared_lib_target_platform_x86_64_linux_test, target_under_test = "//zig/tests/simple-shared-library:library"),
+        partial.make(_build_shared_lib_target_platform_aarch64_linux_test, target_under_test = "//zig/tests/simple-shared-library:library"),
+        partial.make(_build_shared_lib_target_platform_x86_64_windows_test, target_under_test = "//zig/tests/simple-shared-library:library"),
```
For these to work you need to add corresponding `_build_shared_lib_target...` definitions [similar to lines 60-62](https://github.com/aherrmann/rules_zig/pull/103/files#diff-a36ccfd0adf6052d9a186d49b17c840e3e6fff7fafec85bcf07f21a9e06fac13R60-R62).

aherrmann reviewed (commented):
Thank you, that's already looking great! You're close.

One small comment (some test definitions are missing).

> I'm not sure what command to run for the cross-platform tests. @aherrmann please advise.

These tests are included in the standard test suites. So, they'll run when you run `bazel test //...`. To run these tests specifically you can use the command `bazel test //zig/tests:target_platform_test`. (The tests use Bazel's transitions system under the hood to test different target platforms.)

aherrmann commented on zig/private/common/zig_build.bzl:
```diff
@@ -125,9 +125,15 @@ def zig_build_impl(ctx, *, kind):
 
         files = depset([static])
+    elif kind == "zig_shared_library":
+        extension = ".dll" if zigtargetinfo.triple.os == "windows" else ".so"
+        dynamic = ctx.actions.declare_file(ctx.label.name + extension)
+        outputs.append(dynamic)
+        args.add(dynamic, format = "-femit-bin=%s")
```
This produces outputs following the target name directly, e.g. the following produces `library.so`.

```
$ cd e2e/workspace
$ bazel build //simple-shared-library:library
...
Target //simple-shared-library:library up-to-date:
  bazel-bin/simple-shared-library/library.so
...
```

On Linux and MacOS the linker typically expects a `lib` prefix, so in that case the rule should generate an output. I'm not a Windows expert, but, [to my understanding](https://learn.microsoft.com/en-us/cpp/build/walkthrough-creating-and-using-a-dynamic-link-library-cpp?view=msvc-170) Windows does not use the `lib` prefix.

The following change should achieve this:
```suggestion
        prefix = "" if zigtargetinfo.triple.os == "windows" else "lib"
        extension = ".dll" if zigtargetinfo.triple.os == "windows" else ".so"
        dynamic = ctx.actions.declare_file(prefix + ctx.label.name + extension)
        outputs.append(dynamic)
        args.add(dynamic, format = "-femit-bin=%s")
```

aherrmann commented on zig/private/common/zig_build.bzl:
```diff
@@ -195,6 +201,10 @@ def zig_build_impl(ctx, *, kind):
         arguments = ["build-lib", args]
         mnemonic = "ZigBuildLib"
         progress_message = "Building %{input} as Zig library %{output}"
+    elif kind == "zig_shared_library":
+        arguments = ["build-lib", "-dynamic", args]
+        mnemonic = "ZigBuildLib"
+        progress_message = "Building %{input} as Zig shared library %{output}"
```
If I inspect the output of the test described above I see that the soname of the library doesn't match the actual output. This is likely to cause trouble when linking against this library, as the corresponding load commands will use the wrong name.
```
$ readelf -d bazel-bin/simple-shared-library/library.so
...
  Tag        Type                         Name/Value
 0x000000000000000e (SONAME)             Library soname: [libmain.so]
...
```

The Zig compiler accepts a flag to change the SONAME:
```
  -fsoname[=name]                Override the default SONAME value
```

The following change should do the right thing:
```suggestion
    elif kind == "zig_shared_library":
        arguments = ["build-lib", "-dynamic", "-fsoname" + dynamic.basename, args]
        mnemonic = "ZigBuildLib"
        progress_message = "Building %{input} as Zig shared library %{output}"
```

Titousensei commented:
Updated to fix the tests and address the comments:
- renamed library target name to `shared` (was `library`), so make it more obvious the the name `libshared.so` is what's intended
- added prefix for linux and macos
- fixed `soname`
- fixed tests and added more (threaded and mode)

Titousensei commented:
Fixed tests failures related to defs import

aherrmann reviewed (approved):
Thank you for contributing this feature!

I fixed a few more minor issues. I also noticed that there was a typo in the docs, these docs are automatically generated by running `bazel run //docs:update`, or by `bazel run //utils:update` which also updates other generated files.
